### PR TITLE
support configurable retain & base subscription only on mqtt settings

### PR DESCRIPTION
**Pull request by chrisns**
Since some servers throw errors and drop the connections if you try (e.g. aws iot)
Happy to PR the docs if you accept this 😄

### CODE REVIEW

**Review by nurikk — APPROVED** [review]

**Comment by nurikk** [comment]
There are more places in code where messages are published with retain flag
for example
https://github.com/Koenkk/zigbee2mqtt/blob/9ed84e130fa1fd2deb310f72b70387850d4e1388/lib/extension/bridge.js#L503

**Comment by chrisns** [comment]
> There are more places in code where messages are published with retain flag
> for example
> https://github.com/Koenkk/zigbee2mqtt/blob/9ed84e130fa1fd2deb310f72b70387850d4e1388/lib/extension/bridge.js#L503

oh, good spot, hadn't seen that one, i'll move this PR back to WIP and grep the code again

**Comment by Koenkk** [comment]
I don't understand what this PR tries to fix/address, can you explain it?

**Comment by chrisns** [comment]
@Koenkk hey, sorry I should have provided more context.
I'm trying to use zigbee2mqtt with aws's managed MQTT broker (AWS IoT Core) which *really* doesn't like `retain` payloads so much so that it will boot a client for trying.
I believe some other managed brokers behave similar so its not unique to AWS

**Comment by Koenkk** [comment]
I'm asking because having `bridge/state` not set to retain can cause serious issues (e.g. keeping all devices in Home Assistant offline after a Home Assistant restart). Isn't this an issue that should be solved at the AWS side? Having an MQTT broker that does not support retain well looks fundamentally broken to me. I'm wondering if we should support that at all from a Zigbee2MQTT perspective.

**Comment by chrisns** [comment]
its quite a common theme in large scale MQTT brokers see:

 - https://docs.aws.amazon.com/iot/latest/developerguide/mqtt.html#mqtt-differences
 - https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-mqtt-support#sending-device-to-cloud-messages
 - https://cloud.google.com/iot/docs/requirements#mqtt_bridge

They all have different models for handling persistence, I presume its because `retain` has issue with consistency models when you get to cloud vendor scale.

Agree that HA and similar compatibility may vary, my use case of zigbee2mqtt doesn't depend on that

**Comment by Koenkk** [comment]
I see, wasn't aware of this. I think a more sustainable solution is to force disable the retain flag here: https://github.com/Koenkk/zigbee2mqtt/blob/81b8c859af352c9f5b1b8ae0e49848c25b1df094/lib/mqtt.js#L107. This will make sure retain is always disabled, even if in the future other places with retain: true get added. This option should be put under `advaced` and a good name would be `force_disable_retain`.  Please also make an update for the docs and I will accept it.

**Comment by chrisns** [comment]
sure, sounds good. will make changes asap.
tangentially cloud vendors often also limit the subscriptions, at the minute the zigbee2mqtt generates a pretty long list of subscriptions, would you be amenable to another separate PR (or combined if you like?) to allow a configuration option to do broad `prefix/#` rather than deep `prefix/+/+/+/+/etc/action`?

**Comment by chrisns** [comment]
I added the subscribing to the whole base prefix commit, happy to separate it into a different PR conversation if its contentious.
Will make corresponding PR to the docs on https://github.com/Koenkk/zigbee2mqtt.io if this is approved
Happy to wait for them to be merged together if you prefer, but would be good to check your happy with this first :)
Thanks!

File: lib/mqtt.js [hunk 1]
```diff
@@ -74,6 +74,9 @@ class MQTT extends events.EventEmitter {
 
             this.client.on('connect', () => {
                 logger.info('Connected to MQTT server');
+                if (settings.get().advanced.subscribe_to_whole_prefix) {
+                    this.client.subscribe(`${settings.get().mqtt.base_topic}/#`);
```
Comment (Koenkk) [inline lib/mqtt.js:78]: I'm wondering now why we even need to subscribe in all extensions individually. Just subscribing to `zigbee2mqtt/#` here seems like a much easier solution and we can remove all the calls to `subscribe`.

Comment (chrisns) [reply lib/mqtt.js:78]: It does stop you receiving a feedback loop on stuff you publish?

Comment (Koenkk) [reply lib/mqtt.js:78]: I don't understand, I don't expect the cause different behaviour.

Regarding the force disable retain, fixed an issue here that would prevent the frontend from functioning correctly.

Comment (chrisns) [reply lib/mqtt.js:78]: Cool, thanks!

**Comment by Koenkk** [comment]
Merged, thanks!

- Wrote some docs (put it under mqtt instead of advanced): https://github.com/Koenkk/zigbee2mqtt.io/blob/develop/docs/information/configuration.md#configuration
- Simplified the subscribes to only one (if not enabling home assistant integration) (`zigbee2mqtt/#`) in https://github.com/Koenkk/zigbee2mqtt/commit/9c8323326d704dca8f213ef088e5566096efdc7b

**Pull request merged**